void functions in `always_comb` blocks do not contribute to the sensitivity list

I am trying to simulate the following generated SV module along with it's testbench:

**Design: `counter.sv`**

```systemverilog
//==============================================================================
//
// The code is generated by Intel Compiler for SystemC, version 1.3.31
// see more information at https://github.com/intel/systemc-compiler
//
//==============================================================================

//==============================================================================
//
// Module: Dut ()
//
module Dut // "tb.dut_inst"
(
    input logic clk,
    input logic rstn,
    output logic [3:0] counter,
    output logic even
);

//------------------------------------------------------------------------------
// Clocked THREAD: threadProc (dut.h:26:5) 

// Thread-local variables
logic [3:0] counter_next;

// Next-state combinational logic
always_comb begin : threadProc_comb     // dut.h:26:5
    threadProc_func;               //<   Comment this line for iverilog simulation to work
    //counter_next = counter;      //< Uncomment this line for iverilog simulation to work
    //counter_next = counter + 1;  //< Uncomment this line for iverilog simulation to work
end
function void threadProc_func;
    counter_next = counter;
    counter_next = counter + 1;
endfunction

// Syncrhonous register update
always_ff @(posedge clk or negedge rstn) 
begin : threadProc_ff
    if ( ~rstn ) begin
        counter <= 0;
    end
    else begin
        counter <= counter_next;
    end
end

//------------------------------------------------------------------------------
// Method process: methodProc (dut.h:35:5) 

always_comb 
begin : methodProc     // dut.h:35:5
 even = |(counter % 2);
end

endmodule
```
**Design Testbench: `tb_counter.sv`**
```systemverilog
module tb_counter();

logic clk;
logic rstn;
logic [3:0] counter;
logic even;

Dut DUT0(clk, rstn, counter, even);

always
  #10  clk= ~clk;


initial
begin
    rstn = 1;
    clk = 0;
    #20 rstn = 0;

    #30 rstn = 1;

    #500 $finish;
end

initial begin
        if ($test$plusargs("trace") != 0) begin
                $display("[%0t] Tracing to vlt_dump.vcd...\n", $time);
                $dumpfile("logs/vlt_dump.vcd");
                $dumpvars;
        end
        $display("[%0t] Model Running...\n", $time);
end

endmodule
```
Getting the following error while trying out build and run the simulation:

**Iverilog Build command:**
```bash
iverilog -g2012 counter.sv tb_counter.sv -o counter.aout
```
**Output:**
```
counter.sv:27: warning: always_comb process has no sensitivities.
ivl: ../net_scope.cc:500: const NetTaskDef* NetScope::task_def() const: Assertion `type_ == TASK' failed.
```

If one comments the line `28` in `counter.sv` and un-comment the lines `29-30` iverilog is able to extract sensitivity.

Here are the details below for the `iverilog` build from source:

**Iverilog Build info:`iverilog -v`**

```txt
iverilog -v
Icarus Verilog version 12.0 (devel) (s20150603-1147-g7ee7a483)

Copyright (c) 2000-2021 Stephen Williams (steve@icarus.com)

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along
  with this program; if not, write to the Free Software Foundation, Inc.,
  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

iverilog: no source files.

Usage: iverilog [-EiSuvV] [-B base] [-c cmdfile|-f cmdfile]
                [-g1995|-g2001|-g2005|-g2005-sv|-g2009|-g2012] [-g<feature>]
                [-D macro[=defn]] [-I includedir] [-L moduledir]
                [-M [mode=]depfile] [-m module]
                [-N file] [-o filename] [-p flag=value]
                [-s topmodule] [-t target] [-T min|typ|max]
                [-W class] [-y dir] [-Y suf] [-l file] source_file(s)

See the man page for details.
```
Let me know in-case any more details are required.
Seems to be related to #159 .

Looks like Icarus uses always@(*) semantics to extract sensitivity list and
does not support the inference via a function (inside always comb).

Regards
Srini


On Sat, 21 Aug, 2021, 18:53 Ameya Vikram Singh, ***@***.***>
wrote:

> I am trying to simulate the following generated SV module along with it's
> testbench:
>
> *Design: counter.sv <http://counter.sv>*
>
> //==============================================================================//// The code is generated by Intel Compiler for SystemC, version 1.3.31// see more information at https://github.com/intel/systemc-compiler////==============================================================================
> //==============================================================================//// Module: Dut ()//module Dut // "tb.dut_inst"
> (
>     input logic clk,
>     input logic rstn,
>     output logic [3:0] counter,
>     output logic even
> );
> //------------------------------------------------------------------------------// Clocked THREAD: threadProc (dut.h:26:5)
> // Thread-local variableslogic [3:0] counter_next;
> // Next-state combinational logicalways_comb begin : threadProc_comb     // dut.h:26:5
>     threadProc_func;               //<   Comment this line for iverilog simulation to work
>     //counter_next = counter;      //< Uncomment this line for iverilog simulation to work
>     //counter_next = counter + 1;  //< Uncomment this line for iverilog simulation to workendfunction void threadProc_func;
>     counter_next = counter;
>     counter_next = counter + 1;endfunction
> // Syncrhonous register updatealways_ff @(posedge clk or negedge rstn) begin : threadProc_ff
>     if ( ~rstn ) begin
>         counter <= 0;
>     end
>     else begin
>         counter <= counter_next;
>     endend
> //------------------------------------------------------------------------------// Method process: methodProc (dut.h:35:5)
> always_comb begin : methodProc     // dut.h:35:5
>  even = |(counter % 2);end
> endmodule
>
> *Design Testbench: tb_counter.sv <http://tb_counter.sv>*
>
> module tb_counter();
> logic clk;logic rstn;logic [3:0] counter;logic even;
> Dut DUT0(clk, rstn, counter, even);
> always
>   #10  clk= ~clk;
>
> initialbegin
>     rstn = 1;
>     clk = 0;
>     #20 rstn = 0;
>
>     #30 rstn = 1;
>
>     #500 $finish;end
> initial begin
>         if ($test$plusargs("trace") != 0) begin
>                 $display("[%0t] Tracing to vlt_dump.vcd...\n", $time);
>                 $dumpfile("logs/vlt_dump.vcd");
>                 $dumpvars;
>         end
>         $display("[%0t] Model Running...\n", $time);end
> endmodule
>
> Getting the following error while trying out build and run the simulation:
>
> *Iverilog Build command:*
>
> iverilog -g2012 counter.sv tb_counter.sv -o counter.aout
>
> *Output:*
>
> counter.sv:27: warning: always_comb process has no sensitivities.
> ivl: ../net_scope.cc:500: const NetTaskDef* NetScope::task_def() const: Assertion `type_ == TASK' failed.
>
> If one comments the line 28 in counter.sv and un-comment the lines 29-30
> iverilog is able to extract sensitivity.
>
> Here are the details below for the iverilog build from source:
>
> *Iverilog Build info:iverilog -v*
>
> iverilog -v
> Icarus Verilog version 12.0 (devel) (s20150603-1147-g7ee7a483)
>
> Copyright (c) 2000-2021 Stephen Williams ***@***.***)
>
>   This program is free software; you can redistribute it and/or modify
>   it under the terms of the GNU General Public License as published by
>   the Free Software Foundation; either version 2 of the License, or
>   (at your option) any later version.
>
>   This program is distributed in the hope that it will be useful,
>   but WITHOUT ANY WARRANTY; without even the implied warranty of
>   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
>   GNU General Public License for more details.
>
>   You should have received a copy of the GNU General Public License along
>   with this program; if not, write to the Free Software Foundation, Inc.,
>   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
>
> iverilog: no source files.
>
> Usage: iverilog [-EiSuvV] [-B base] [-c cmdfile|-f cmdfile]
>                 [-g1995|-g2001|-g2005|-g2005-sv|-g2009|-g2012] [-g<feature>]
>                 [-D macro[=defn]] [-I includedir] [-L moduledir]
>                 [-M [mode=]depfile] [-m module]
>                 [-N file] [-o filename] [-p flag=value]
>                 [-s topmodule] [-t target] [-T min|typ|max]
>                 [-W class] [-y dir] [-Y suf] [-l file] source_file(s)
>
> See the man page for details.
>
> Let me know in-case any more details are required.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/steveicarus/iverilog/issues/539>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADEBKW77226JEAUPVVQ3HQDT56SE7ANCNFSM5CR3G3CQ>
> .
>

--- a/ivtest/ivltests/always_comb_void_func.v
+++ b/ivtest/ivltests/always_comb_void_func.v
@@ -1,0 +1,51 @@
+// Check that variables referenced in a void function contribute to the
+// sensitivity list of a always_comb block.
+
+module top;
+  logic passed;
+  logic [7:0] value;
+  integer counter;
+
+  function automatic void count(bit what);
+    counter = 0;
+    for (integer i = 0; i < $bits(value); i++) begin
+      if (value[i] == what)
+        counter += 1;
+    end
+  endfunction
+
+  always_comb begin
+    count(1'b1);
+  end
+
+  initial begin
+    passed = 1'b1;
+
+    value = 8'b0000_0000;
+    #1;
+    if (counter !== 0) begin
+      $display("Expected 0, got %d", counter);
+      passed = 1'b0;
+    end
+
+    value = 8'b0011_1100;
+    #1;
+    if (counter !== 4) begin
+      $display("Expected 4, got %d", counter);
+      passed = 1'b0;
+    end
+
+    value = 8'b1011_1101;
+    #1;
+    if (counter !== 6) begin
+      $display("Expected 6, got %d", counter);
+      passed = 1'b0;
+    end
+
+    if (passed) begin
+      $display("PASSED");
+    end else begin
+      $display("FAILED");
+    end
+  end
+endmodule

--- a/ivtest/regress-sv.list
+++ b/ivtest/regress-sv.list
@@ -76,6 +76,7 @@ always_comb_fail4	CE,-g2005-sv		ivltests
 always_comb_no_sens	nornal,-g2005-sv	ivltests gold=always_comb_no_sens.gold
 always_comb_rfunc	nornal,-g2005-sv	ivltests
 always_comb_trig	normal,-g2005-sv	ivltests
+always_comb_void_func	normal,-g2005-sv	ivltests
 always_comb_warn	normal,-g2005-sv	ivltests gold=always_comb_warn.gold
 always_ff		normal,-g2005-sv	ivltests
 always_ff_fail		CE,-g2005-sv		ivltests

--- a/ivtest/regress-vlog95.list
+++ b/ivtest/regress-vlog95.list
@@ -69,6 +69,7 @@
 
 # Verilog 95 does not support automatic tasks or functions.
 always_comb_rfunc	CE			ivltests
+always_comb_void_func	CE			ivltests
 automatic_error11	CE			ivltests
 automatic_error12	CE			ivltests
 automatic_error13	CE			ivltests

--- a/net_nex_input.cc
+++ b/net_nex_input.cc
@@ -255,6 +255,30 @@ NexusSet* NetETernary::nex_input(bool rem_out, bool always_sens, bool nested_fun
       return result;
 }
 
+// Get the contribution of a function call in a always_comb block
+static void func_always_sens(NetFuncDef *func, NexusSet *result,
+			     bool rem_out, bool nested_func)
+{
+	  // Avoid recursive function calls.
+	static set<NetFuncDef*> func_set;
+	if (!nested_func)
+	      func_set.clear();
+
+	if (!func_set.insert(func).second)
+	      return;
+
+	std::unique_ptr<NexusSet> tmp(func->proc()->nex_input(rem_out, true, true));
+	  // Remove the function inputs
+	std::unique_ptr<NexusSet> in(new NexusSet);
+	for (unsigned idx = 0; idx < func->port_count(); idx++) {
+	      NetNet *net = func->port(idx);
+	      assert(net->pin_count() == 1);
+	      in->add(net->pin(0).nexus(), 0, net->vector_width());
+	}
+	tmp->rem(*in);
+	result->add(*tmp);
+}
+
 NexusSet* NetEUFunc::nex_input(bool rem_out, bool always_sens, bool nested_func) const
 {
       NexusSet*result = new NexusSet;
@@ -265,31 +289,8 @@ NexusSet* NetEUFunc::nex_input(bool rem_out, bool always_sens, bool nested_func)
 	    delete tmp;
       }
 
-      if (always_sens) {
-	    NetFuncDef*func = func_->func_def();
-
-	      // Avoid recursive function calls.
-	    static set<NetFuncDef*> func_set;
-	    if (!nested_func)
-		  func_set.clear();
-
-	    if (!func_set.insert(func).second)
-		  return result;
-
-	    NexusSet*tmp = func->proc()->nex_input(rem_out, always_sens, true);
-	      // Remove the function inputs
-	    NexusSet*in = new NexusSet;
-	    for (unsigned idx = 0 ;  idx < func->port_count() ;  idx += 1) {
-		  NetNet*net = func->port(idx);
-		  assert(net->pin_count() == 1);
-		  in->add(net->pin(0).nexus(), 0, net->vector_width());
-	    }
-	    tmp->rem(*in);
-	    delete in;
-
-	    result->add(*tmp);
-	    delete tmp;
-      }
+      if (always_sens)
+	    func_always_sens(func_->func_def(), result, rem_out, nested_func);
 
       return result;
 }
@@ -613,9 +614,18 @@ NexusSet* NetSTask::nex_input(bool rem_out, bool always_sens, bool nested_func) 
  * parameters to consider, because the compiler already removed them
  * and converted them to blocking assignments.
  */
-NexusSet* NetUTask::nex_input(bool, bool, bool) const
+NexusSet* NetUTask::nex_input(bool rem_out, bool always_sens, bool nested_func) const
 {
-      return new NexusSet;
+      NexusSet *result = new NexusSet;
+
+      /*
+       * Let the contents of void functions contribute to the sensitivity list
+       * of always_comb blocks
+       */
+      if (always_sens && task_->type() == NetScope::FUNC)
+	    func_always_sens(task_->func_def(), result, rem_out, nested_func);
+
+      return result;
 }
 
 NexusSet* NetWhile::nex_input(bool rem_out, bool always_sens, bool nested_func) const

